src/icelake/icelake_utf8_length_from_utf16.inl.cpp

```diff
@@ -1,84 +1,294 @@
-// This is translation of `utf8_length_from_utf16_bytemask` from
-// `generic/utf16.h`
 template <endianness big_endian>
 simdutf_really_inline size_t icelake_utf8_length_from_utf16(const char16_t *in,
                                                             size_t size) {
-  size_t pos = 0;
 
   using vector_u16 = simd16<uint16_t>;
-  constexpr size_t N = vector_u16::ELEMENTS;
-
-  const auto one = vector_u16::splat(1);
+  constexpr size_t N = vector_u16::ELEMENTS; // 32 on AVX-512
+  if (N + 1 > size) {
+    return scalar::utf16::utf8_length_from_utf16_with_replacement<big_endian>(
+        in, size);
+  } // special case for short inputs
+  size_t pos = 0;
 
-  auto v_count = vector_u16::zero();
+  const __m512i byteflip = _mm512_setr_epi64(
+      0x0607040502030001, 0x0e0f0c0d0a0b0809, 0x0607040502030001,
+      0x0e0f0c0d0a0b0809, 0x0607040502030001, 0x0e0f0c0d0a0b0809,
+      0x0607040502030001, 0x0e0f0c0d0a0b0809);
 
   // each char16 yields at least one byte
-  size_t count = size / N * N;
+  size_t count = 0;
 
-  // in a single iteration the increment is 0, 1 or 2, despite we have
-  // three additions
-  constexpr size_t max_iterations = 65535 / 2;
-  size_t iteration = max_iterations;
+  for (; pos < size / (2 * N) * (2 * N); pos += 2 * N) {
```

Comment:
why this particular iteration size? also maybe better if we assign this to a variable

Reply:
We unroll by a factor of two for performance. We might be able to squeeze more performance by unrolling further, but we are likely to hit diminishing returns. I don't think we can safely change this range here without changing the rest of the code, so it is not codes as a parameter. We could do so, of course. But I expect that it is maybe not a place where we have much to gain. We already cover 128 bytes per iteration.